workflow/controller/controller.go

```diff
@@ -338,6 +330,15 @@ func (wfc *WorkflowController) Run(ctx context.Context, wfWorkers, workflowTTLWo
 		log.Fatal("Timed out waiting for caches to sync")
 	}
 
+	// init managers: throttler and SynchronizationManager
+	if err := wfc.initManagers(); err != nil {
+		log.Fatal(err)
+	}
+
+	if err := wfc.addWorkflowInformerHandlers(ctx); err != nil {
```

Comment:
I am concerned about the event handlers being added _after_ the Informer has started and synced. That makes me think that some events will be missed and that the informer will not be properly filtered (resulting in higher memory usage).

I'm thinking that the approach I mentioned in my second paragraph of https://github.com/argoproj/argo-workflows/pull/13304/files#r1666940943 would be more straightforward

Reply:
If I understand correctly, you mean to sequentially process adding the workflows to the syncManager and throttler in the `AddFunc` of the informerHandler, and then proceed with the original AddFunc logic. This way, the order of addition can be ensured. However, it can only be processed one at a time.

Reply:
I have another question, if we add the workflows to be processed into the manager within the informer, how do we determine which ones need to be added? Since the event handler is continuously running.

Reply:
Trying to follow all of this...

> I am concerned about the event handlers being added after the Informer has started and synced. That makes me think that some events may be missed and that the informer may not be properly filtered (resulting in higher memory usage).

So, the danger here is that the FilterFunc is being applied to the cache after the call to `wfc.initManagers()`? In that case, the danger to me seems to be related to: _what happens if that function processes Workflows which are in a "Completed" state?_  I think I'm missing the part about higher memory usage. As for memory usage, it seems to me that given that `NewWorkflowInformer()` has always been called prior to `addWorkflowInformerHandlers()` which includes the `FilterFunc`, that there has always been some temporary usage of memory that has included all of those Completed Workflows, which would actually risk Controller getting into an OOM crash loop ? (anyway, we don't need to worry about that here)

Am I correct that `addWorkflowInformerHandlers()` was moved to after `initManagers()` because the `AddFunc()` uses the throttler, which must be initialized first? In that case, it seems that the WorkflowInformer must be created with its `FilterFunc`. Then `initManagers()` should be called. Then the Workflow Informer Handlers can be created?

Sorry if I'm off. I also wasn't sure what @agilgur5 's suggestion was.